### PR TITLE
[IMPROVEMENT] Migrate AsyncStorage to RealmDB

### LINKED ISSUE #919: Migrate AsyncStorage to RealmDB
Asyncstorage complicates the code because it creates one more data access point and does not accept objects, which creates unnecessary castings, migrate all data to the realm creating a table of users and of server could improve this.

### PR DESCRIPTION
**Pull request by djorkaeffalexandre**
Using AsyncStorage to store data at the same time as RealmDB can add complexity to the code and so migrating user login data to realmDB may be a good option.

The use of asyncstorage is impeditive for App Groups, it is not possible to access the asyncstorage data in other app containers that may be in eg. Share Extension, so I migrated the data from asyncstorage to realmdb by associating the logged in user with the server table and a variable to indicate the current server.

Closes #919

### CODE REVIEW

File: app/sagas/init.js [hunk 1]
```diff
@@ -14,9 +13,13 @@ import database from '../lib/realm';
 
 const restore = function* restore() {
 	try {
+		const { serversDB } = database.databases;
+
+		const currentServer = serversDB.objects('servers').filtered('currentServer = true');
```
Comment (IlarionHalushka) [inline app/sagas/init.js]: do `const { userToken: token, id: server } = yield serversDB.objects('servers').filtered('currentServer = true')[0];` here and remove next lines where you are defining `{token, server}`

File: app/sagas/deepLinking.js [hunk 2]
```diff
@@ -42,9 +41,12 @@ const handleOpen = function* handleOpen({ params }) {
 		host = host.slice(0, host.length - 1);
 	}
 
+	const { serversDB } = database.databases;
+	const currentServer = serversDB.objects('servers').filtered('currentServer = true');
```
Comment (IlarionHalushka) [inline app/sagas/deepLinking.js]: do `const { id: server, user } = yield serversDB.objects('servers').filtered('currentServer = true')[0];` and remove next lines where you are defining `const [server, user] =`

File: app/lib/rocketchat.js [hunk 3]
```diff
@@ -74,7 +74,8 @@ const RocketChat = {
 
 	async getUserToken() {
 		try {
-			return await AsyncStorage.getItem(TOKEN_KEY);
+			const { serversDB } = database.databases;
+			return await serversDB.objects('servers').filtered('currentServer = true')[0].user;
```
Comment (IlarionHalushka) [inline app/lib/rocketchat.js]: `return serversDB.objects('servers').filtered('currentServer = true')[0].user` no need for await when you are returning from function

File: app/lib/realm.js [hunk 4]
```diff
@@ -1,15 +1,39 @@
 import Realm from 'realm';
 
-// import { AsyncStorage } from 'react-native';
-// Realm.clearTestState();
-// AsyncStorage.clear();
+const email = {
+	name: 'email',
+	primaryKey: 'address',
+	properties: {
+		// _id: { type: 'int', optional: true },
```
Comment (IlarionHalushka) [inline app/lib/realm.js]: ?

File: app/lib/rocketchat.js [hunk 5]
```diff
@@ -74,7 +74,8 @@ const RocketChat = {
...
 		} catch (error) {
 			console.warn(`AsyncStorage error: ${ error.message }`);
```
Comment (IlarionHalushka) [inline app/lib/rocketchat.js]: update console warnings where `AsyncStorage` is mentioned

File: app/sagas/init.js [hunk 6]
```diff
@@ -26,10 +29,6 @@ const restore = function* restore() {
 		yield put(toggleMarkdown(useMarkdown));
 
 		if (!token || !server) {
-			yield all([
```
Comment (IlarionHalushka) [inline app/sagas/init.js]: why no equivalent for removing current server and token using realm?

File: app/lib/realm.js [hunk 7]
```diff
@@ -1,15 +1,39 @@
...
+};
 
 const serversSchema = {
 	name: 'servers',
 	primaryKey: 'id',
 	properties: {
 		id: 'string',
 		name: { type: 'string', optional: true },
+		currentServer: { type: 'bool', optional: true },
+		userToken: { type: 'string', optional: true },
```
Comment (IlarionHalushka) [inline app/lib/realm.js]: you already declared `token` field in `user` schema on line 17 same file. So you can access `servers.userToken` using `servers.user.token. I think no need in duplicated token field value in two places.

File: app/lib/rocketchat.js (hunk 5 above)
Comment (IlarionHalushka) [reply app/lib/rocketchat.js]: would be nice of you to use `log()` :)

File: app/sagas/login.js [hunk 8]
```diff
@@ -65,13 +72,13 @@ const handleLogout = function* handleLogout() {
 			const servers = yield serversDB.objects('servers');
 			// filter logging out server and delete it
 			const serverRecord = servers.filtered('id = $0', server);
+			const token = serverRecord[0].user;
```
Comment (IlarionHalushka) [inline app/sagas/login.js]: define it inside of `if (servers.length > 0) {` (line 80 same file)

File: app/sagas/init.js (hunk 6 above)
Comment (djorkaeffalexandre) [reply app/sagas/init.js]: because if ```.objects('servers').filtered('currentServer = true')``` returns ```null``` haven't servers to remove.

**Comment by djorkaeffalexandre** [comment]
> Please describe how you changed realm structure and what is missing.

The use of asyncstorage is impeditive for App Groups, it is not possible to access the asyncstorage data in other app containers that may be in eg. Share Extension, so I migrated the data from asyncstorage to realmdb by associating the logged in user with the server table and a variable to indicate the current server.

**Comment by djorkaeffalexandre** [comment]
I need to do a migration to get AsyncStorage old data and put it on RealmDB servers schema.

File: app/sagas/selectServer.js [hunk 9]
```diff
@@ -36,11 +36,26 @@ const handleSelectServer = function* handleSelectServer({ server, version, fetch
...
-		if (userStringified) {
-			const user = JSON.parse(userStringified);
+		const { serversDB } = database.databases;
+		const servers = yield serversDB.objects('servers');
+
+		serversDB.write(() => {
+			try {
+				// eslint-disable-next-line no-plusplus
+				for (let i = 0; i < servers.length; i++) {
+					serversDB.create('servers', { id: servers[i].id, currentServer: false }, true);
```
Comment (IlarionHalushka) [inline app/sagas/selectServer.js]: `serversDB.create('servers', { id: servers[i].id, currentServer: servers[i].id === server }, true);` and then no need to make additional request to DB on line next line: `49`

File: app/views/SidebarView/index.js [hunk 10]
```diff
@@ -153,7 +153,7 @@ export default class Sidebar extends Component {
 			const permissionsFiltered = database.objects('permissions')
 				.filter(permission => permissions.includes(permission._id));
 			return permissionsFiltered.reduce((result, permission) => (
-				result || permission.roles.some(r => roles.includes(r))),
+				result || permission.roles.some(r => roles.indexOf(r) !== -1)),
```
Comment (IlarionHalushka) [inline app/views/SidebarView/index.js]: why?

File: app/lib/realm.js [hunk 11]
```diff
@@ -1,15 +1,39 @@
 import Realm from 'realm';
-
 // import { AsyncStorage } from 'react-native';
-// Realm.clearTestState();
-// AsyncStorage.clear();
+// import RocketChat from './rocketchat';
```
Comment (diegolmello) [inline app/lib/realm.js]: Add this line. 
```
// Realm.clearTestState();
```

It's useful during development and it's deleted in build release anyway.

File: app/lib/realm.js [hunk 12]
```diff
@@ -1,15 +1,39 @@
...
+	name: 'user',
+	properties: {
+		id: { type: 'string', optional: true },
+		token: { type: 'string', optional: true },
+		username: { type: 'string', optional: true },
+		name: { type: 'string', optional: true },
+		language: { type: 'string', optional: true },
+		status: { type: 'string', optional: true },
+		customFields: { type: 'string', optional: true },
+		email: { type: 'list', objectType: 'email', default: [] },
```
Comment (diegolmello) [inline app/lib/realm.js]: It seem to be a typo.
Callback from `api/v1/login` is `emails`.

File: app/lib/realm.js [hunk 13]
```diff
@@ -1,15 +1,39 @@
...
+const userSchema = {
+	name: 'user',
+	properties: {
+		id: { type: 'string', optional: true },
+		token: { type: 'string', optional: true },
+		username: { type: 'string', optional: true },
+		name: { type: 'string', optional: true },
+		language: { type: 'string', optional: true },
+		status: { type: 'string', optional: true },
+		customFields: { type: 'string', optional: true },
```
Comment (diegolmello) [inline app/lib/realm.js]: This must be an object.

File: app/lib/realm.js [hunk 14]
```diff
@@ -1,15 +1,39 @@
 import Realm from 'realm';
-
 // import { AsyncStorage } from 'react-native';
-// Realm.clearTestState();
-// AsyncStorage.clear();
+// import RocketChat from './rocketchat';
+
+const email = {
```
Comment (diegolmello) [inline app/lib/realm.js]: Call it `userEmailSchema`

File: app/sagas/login.js [hunk 15]
```diff
@@ -33,15 +32,23 @@ const handleLoginRequest = function* handleLoginRequest({ credentials }) {
...
+		serversDB.write(() => {
+			try {
+				serversDB.create('servers', { id: server, userToken: user.token, user }, true);
+			} catch (e) {
+				log('handleLoginSuccess -> setUserToken ->', e);
+			}
+		});
 	} catch (error) {
-		console.log('loginSuccess saga -> error', error);
+		log('loginSuccess saga -> error', error);
```
Comment (diegolmello) [inline app/sagas/login.js]: Just merged a PR with new `log` names. Use snake case.

File: app/sagas/login.js [hunk 16]
```diff
@@ -59,19 +66,19 @@ const handleLogout = function* handleLogout() {
...
 	if (server) {
 		try {
-			yield call(logoutCall, { server });
+			yield call(logoutCall);
 			const { serversDB } = database.databases;
 			// all servers
 			const servers = yield serversDB.objects('servers');
 			// filter logging out server and delete it
 			const serverRecord = servers.filtered('id = $0', server);
+			const token = serverRecord[0].user;
```
Comment (diegolmello) [inline app/sagas/login.js]: It'll crash if `serverRecord.length === 0`;

File: app/views/SidebarView/index.js (hunk 10 above)
Comment (djorkaeffalexandre) [reply app/views/SidebarView/index.js]: I don't know but ```includes``` cause an error, I tried to change this based on this issue:
https://github.com/facebook/create-react-app/issues/942

File: app/sagas/selectServer.js (hunk 9 above)
Comment (djorkaeffalexandre) [reply app/sagas/selectServer.js]: nice

**Comment by diegolmello** [comment]
As talked offline, it'll be better if we use App Groups instead of realm.
I'm closing this one.

**Pull request closed without merging**